### ISSUE TITLE
Bulk Export [CI] - Export contains some duplicate files

### ISSUE THREAD

**Issue by spetillo** [WHOIGit/ooicgsn-roundabout#148]
See screen shot. Some files are exported with duplicate names (see files that have " (1)" in the name).

This isn't really breaking anything, so the fix doesn't need to hold up v1.5.0 release.

**Comment by spetillo** [comment]
![Screenshot from 2020-08-29 18-57-23](https://user-images.githubusercontent.com/11924613/91647535-d79a6500-ea29-11ea-989c-3eaf0655f211.png)

**Comment by sbatchelder** [comment]
Yes I think this is a duplicate data in the database issue. I noticed multiple same-name files during development which is why I put in the feature to annotate duplicate DB entries (calibration events for same instrument in the same day) with a number and perens (1).

**Comment by spetillo** [comment]
Ok, so that's an intentional implementation. Makes sense. I dug in a bit further, and it's clearer now. If you do a new Bulk Export [CI] right now and look at CGINS-CTDBPF-50001__20200830*.csv files, you will see that:
CGINS-CTDBPF-50001__20200830.csv has Cals + _all_ Configs dated 20200830 (not just the ones marked "Export with CSV")
CGINS-CTDBPF-50001__20200830 (1).csv has Cals + _all_ Constants dated 20200830 (not just the ones marked "Export with CSV")
CGINS-CTDBPF-50001__20200830 (2).csv has Cals + _all_ Configs dates "TBD" (not just the ones marked "Export with CSV")

We're getting close! There are just a few tweaks that need to be made:

- [ ] 1) All configs & constants are exported, not just the ones marked "Export with CSV". We should only be exporting the ones marked "Export with CSV".

- [ ] 2) Configs/Constants dated TBD should not be included in the export.

- [ ] 3) Configs and Constants that are marked for "Export with CSV" should be combined into the same file with the Cals being exported. So, this Bulk Export should only give me **one** CSV for CGINS-CTDBPF-50001__20200830. The contents should be:

> serial,name,value,notes
> 16-50001,cal3,555,
> 16-50001,config2,abc,
> 16-50001,hi2,2,

If we get this far for v1.5.0, I will be happy. We can punt the rest to v1.5.1, since I expect it will need some further logic clarification.

- [ ] 4) Finally, to complicate things further, only Cal/Const/Config Events that are _all_ marked as _Approved_ should be exported for CI. Since I _haven't_ actually approved the Constant Event for 8/30/20 yet AND one of the Constant Names is marked as "Export with CSV", then _no_ CGINS-CTDBPF-50001__20200830.csv file should be exported yet.
Basically, we should be treating all "Export with CSV" Configs & Constants as _required_ fields for CSV export, and the Config and Constant Events that get associated with each other need to have the same date.

**Comment by spetillo** [comment]
(Just noting that this was spun out of initial work done in ticket #119 )

**Comment by sbatchelder** [comment]
I think I figured out how to address the bug. 
To make things less confusing:
> A = CalibrationEvent
> B = ConfigEvent (type=configs)
> C = ConfigEvent (type=constants)

The Rules for grouping CSVs are:

1. C's are carried forward until the next C, but do not trigger csv writes. Only "Export with CSV" values are included.
2. A's are carried forward until the next A, but if another A is encountered before a B, a csv write is triggered for the first A (incl. C if previously encountered).
3. B's trigger csv writes, including values from any previously encountered A's and C's. From B, only "Export with CSV" values are included. 
4. B's are skipped if they do not have an associated deployment ("TBD" status)
5. if a resultant file were to be empty, skip it.

I'm going to use CGINS-CTDBPF-0001 (as per backup_2020_09_02T05_15_01.sql.gz from staging).
The list of calibs, configs, constants for this CTDBPF, ordered by date, are as follows, where:

> A<sub>1</sub> A<sub>2</sub> C<sub>1</sub> A<sub>3</sub> B<sub>1</sub> C<sub>2</sub> B<sub>2</sub> 

Note:
- A<sub>1</sub> is empty (no calib values)
- B<sub>2</sub> is "TBD" since it is not associated with a deployment
- A<sub>3</sub> B<sub>1</sub> C<sub>2</sub> B<sub>2</sub> all have the same date: 2020-08-30

Rules 1-3 give us the following groupings:
A<sub>1</sub>
A<sub>1</sub> C<sub>1</sub>
B<sub>1</sub> A<sub>3</sub> C<sub>1</sub>
B<sub>2</sub> A<sub>3</sub> C<sub>2</sub>

Applying rules 4 and 5 respectively:
B<sub>2</sub> is "TBD" so B<sub>2</sub>A<sub>3</sub>C<sub>2</sub> is skipped.
A<sub>1</sub> is empty, so it is skipped

Final Export Results:
**CGINS-CTDBPF-50001__20200806.csv  <-- A<sub>1</sub>C<sub>1</sub>**
![image](https://user-images.githubusercontent.com/44208509/92034459-4255e400-ed3b-11ea-9390-4ba18715e3c9.png)

**CGINS-CTDBPF-50001__20200830.csv  <-- B<sub>1</sub>A<sub>3</sub>C<sub>1</sub>**
![image](https://user-images.githubusercontent.com/44208509/92034559-674a5700-ed3b-11ea-8d96-45dd868cad40.png)


**NEW BUG?**
B<sub>1</sub>A<sub>3</sub>C<sub>1</sub> should probably ACTUALLY be B<sub>1</sub>A<sub>3</sub>C<sub>**2**</sub>
This particular bug is due to C<sub>2</sub> being recorded at the same date as B<sub>1</sub>, making the order they are encountered by the Rules ambiguous. A bugfix for this would be to sort same-day ABC's in "A-C-B" order.

**Comment by sbatchelder** [comment]
Fixed the New bug.
New Final Export:


**CGINS-CTDBPF-50001__20200806.csv <-- A<sub>1</sub>C<sub>1</sub>**   (no change)
![image](https://user-images.githubusercontent.com/44208509/92039408-01fa6400-ed43-11ea-80d1-8db3e738724e.png)

**CGINS-CTDBPF-50001__20200830.csv <-- B<sub>1</sub>A<sub>3</sub>C<sub>2</sub>**   Note hi2 change (from C<sub>1</sub> to C<sub>2</sub>)
![image](https://user-images.githubusercontent.com/44208509/92039347-e98a4980-ed42-11ea-92be-d384c465a212.png)

**Comment by sbatchelder** [comment]
Added "only Cal/Const/Config Events that are all marked as Approved should be exported for CI"
If any of the CCC's  that would be used to output a csv are not approved, the csv is not output.
This feature applies ONLY to Bulk "Export Calibrations with Configs" and "Export All [CI]" exports.

**Comment by spetillo** [comment]
> 1. C's are carried forward until the next C, but do not trigger csv writes. Only "Export with CSV" values are included.

Update: C's can trigger CSV writes, just like B's. So, in the case where we have A2 C1, the file should be dated by C1, not A2.

In real implementations, there should not be multiple B's per event date for a given inventory item, so B1 & B2 should not share the same date (this is not something you should care about in the Export logic, but just something that will be the case with real data in the system).

**Comment by sbatchelder** [comment]
Alright I think I've made the code correction. Before I commit I want to make sure it's what you expect. Here are the updated rules:

The **Rules**:

1. A's are carried forward until the next A, but if another A is encountered before a B _or C_, a csv write is triggered for the first A. Previously recorded C's are included, previously recorded B's are not.
2. B's trigger csv writes, including values from any previously encountered A's and C's. Only "Export with CSV" values are included. B's are skipped if they do not have an associated deployment ("TBD" status)
3. C's trigger csv writes, including values from any previously encountered A's and B's. Only "Export with CSV" values are included. 
4. If a resultant file were to be empty, skip it.
5. CSV files are dated by the Event that triggers the csv write. If there are same-date B and C's in the sequence (both are triggering events), only one csv bearing both B and C is recorded (including any previous A if applicable).

Reference:
> A = CalibrationEvent
> B = ConfigEvent (type=configs)
> C = ConfigEvent (type=constants)

Let's use the following scenario (modified from the list above) where I added A<sub>4</sub> and A<sub>5</sub>.

A<sub>1</sub> A<sub>2</sub> C<sub>1</sub> A<sub>3</sub> **A<sub>4</sub>** B<sub>1</sub> C<sub>2</sub> B<sub>2</sub> **A<sub>5</sub>**

In this example, ABC's (where applicable) are all Approved, non-TBD, have "Export with CSV" flags, and are on unique consecutive dates.
Given the Event Sequence for this hypothetical instrument, are the following Event-bundles below correct, where each bullet represents a csv file _dated_ by the _first_ ABC in the bullet-list.

- A<sub>1</sub>
- C<sub></sub>1 A<sub>2</sub> 
- A<sub>3</sub> C<sub>1</sub> 
- B<sub>1</sub> A<sub>4</sub> 
- C<sub>2</sub> B<sub>1</sub> A<sub>4</sub> 
- B<sub>2</sub> C<sub>2</sub> A<sub>4</sub>

Finally, is it correct then that the trailing **A<sub>5</sub>** is not included in any of the csv's? Not including it is in accordance with the Rules above.

If that's not correct and A<sub>5</sub> __should__ be included in a csv to be downloaded, what would be the correct export formulation that includes it.

**Comment by sbatchelder** [comment]
For testing this without the download buttons on the interface you can use:
`<baseurl>/export/calibration_events_with_configs` or `<baseurl>/export/CI`

Though I guess I should add those export buttons back in now for 1.5.1 and include them in this commit huh

**Comment by ethanandrews** [comment]
@rhmunier @spetillo these updates are now up on the rdb-testing.whoi.edu server.

**Comment by sbatchelder** [comment]
@spetillo and I had a chat this morning and worked out a better way to approach the problem from. 
To summarize, ConfigEvents (which include Constants and Configurations, aka B's and C's) will be exported in associated-deployment pairs, including any previously available Calibrations (A's) if any. 

Furthermore, to be valid a collection of exportable events, they must match (have values for all of) the CCC fields from the inventory's part template. 

**Clarification request:**
Lets say we have A<sub>1</sub> A<sub>2</sub> B<sub>1</sub> C<sub>1</sub> where A<sub>2</sub> is invalid (not approved or doesn't fulfill the expected fields from the part template).
Should the export (a) fail, or (b) contain A<sub>1</sub> B<sub>1</sub> C<sub>1</sub> ?

**Comment by spetillo** [comment]
Hmmm, I'd say (a) fail.

**Comment by sbatchelder** [comment]
Hey! So I think I have the feature working. It was hard to test because all of the available exportables failed, but for valid reasons! The only success/exception is the one test object I made (the "wozzle").

`MODE` is determined by the _part template's_ registered CCC fields.
`<inv>.assembly_part.part` is the only way that I've found to successfully access a part's CCC fields. Some inv's don't have an associated assembly_part so they get skipped. I'm not sure if this is desired behavior.

Here's a terminal readout of my export routine as applied to the DB instance I have.

```
Inv: CGINS-CTDBPF-50001 (id=1089)
Inv calibs: <QuerySet [<CalibrationEvent: 08/30/2020>, <CalibrationEvent: 08/06/2020>, <CalibrationEvent: 08/04/2020>]>
Inv confconst {32: <QuerySet [<ConfigEvent: 08/11/2020>]>, 33: <QuerySet [<ConfigEvent: 08/30/2020>, <ConfigEvent: 08/30/2020>]>}
Part: CTDBP-F (id=811)
Part calib fields     (REQUIRED): {'cal12345', 'cal1'}
Part confconst fields (REQUIRED): {'hi2', 'config2'}
MODE: "part has confconsts"
  Deployment: "12 - Coastal Pioneer" (id=32)
    conf/consts approved:  FAIL. Approvals={<ConfigEvent: 08/11/2020>: False}
    conf/const validation: FAIL. Present={'hi2', 'hi1'} Missing={'config2'}
    calib selected: <CalibrationEvent: 08/06/2020>
    calib approved:   FAIL
    calib validation: PASS
  Deployment: "D11111 - Global Irminger Sea" (id=33)
    conf/consts approved:  FAIL. Approvals={<ConfigEvent: 08/30/2020>: True, <ConfigEvent: 08/30/2020>: False}
    conf/const validation: PASS
    calib selected: <CalibrationEvent: 08/30/2020>
    calib approved:   FAIL
    calib validation: PASS

Inv: CGINS-OPTAAD-00256 (id=1996)
Inv calibs: <QuerySet [<CalibrationEvent: 06/24/2020>]>
Inv confconst {}
FAIL: CGINS-OPTAAD-00256 does not have an associated assembly_part!

Inv: 3750-00169-00002-20001 (id=2451)
Inv calibs: <QuerySet [<CalibrationEvent: 06/01/2020>]>
Inv confconst {27: <QuerySet [<ConfigEvent: 08/19/2020>]>, 28: <QuerySet [<ConfigEvent: 08/19/2020>]>}
Part: VICS (id=922)
Part calib fields     (REQUIRED): {}
Part confconst fields (REQUIRED): {'const3', 'const4', 'STILL TRYING', 'CONFIG5', 'CONFIG4', 'UPDATE2', 'UPDATE', 'CONFIG3'}
MODE: "part has confconsts"
  Deployment: "7 - Testing At Sea" (id=27)
    conf/consts approved:  FAIL. Approvals={<ConfigEvent: 08/19/2020>: False}
    conf/const validation: FAIL. Present={'const3', 'const4'} Missing={'STILL TRYING', 'CONFIG5', 'CONFIG4', 'UPDATE2', 'UPDATE', 'CONFIG3'}
    WARNING: part does not specify any Calibs but inventory contains some. They will be ignored.
    calib approved/validation: N/A 
  Deployment: "4 - Testing At Sea" (id=28)
    conf/consts approved:  FAIL. Approvals={<ConfigEvent: 08/19/2020>: False}
    conf/const validation: FAIL. Present={'const3', 'const4'} Missing={'STILL TRYING', 'CONFIG5', 'CONFIG4', 'UPDATE2', 'UPDATE', 'CONFIG3'}
    WARNING: part does not specify any Calibs but inventory contains some. They will be ignored.
    calib approved/validation: N/A 

Inv: CGINS-CTDBPC-50006 (id=2452)
Inv calibs: <QuerySet [<CalibrationEvent: 06/19/2020>, <CalibrationEvent: 05/19/2018>, <CalibrationEvent: 06/23/2016>]>
Inv confconst {}
Part: CTDBP-C (id=757)
Part calib fields     (REQUIRED): {'cal2', 'New2', 'New1', 'cal1'}
Part confconst fields (REQUIRED): {'new1'}
MODE: "part has confconsts"
  FAIL: inventory confconst is empty

Inv: CGINS-DOSTAD-00134 (id=2453)
Inv calibs: <QuerySet [<CalibrationEvent: 06/23/2020>, <CalibrationEvent: 11/14/2018>]>
Inv confconst {}
FAIL: CGINS-DOSTAD-00134 does not have an associated assembly_part!

Inv: CGINS-OPTAAC-00266 (id=2454)
Inv calibs: <QuerySet [<CalibrationEvent: 06/19/2020>, <CalibrationEvent: 01/09/2020>, <CalibrationEvent: 01/09/2020>, <CalibrationEvent: 08/19/2017>]>
Inv confconst {}
FAIL: CGINS-OPTAAC-00266 does not have an associated assembly_part!

Inv: CGINS-OPTAAD-00124 (id=2651)
Inv calibs: <QuerySet [<CalibrationEvent: 06/26/2020>, <CalibrationEvent: 07/03/2019>]>
Inv confconst {}
FAIL: CGINS-OPTAAD-00124 does not have an associated assembly_part!

Inv: CGINS-CTDMOQ-08523 (id=1436)
Inv calibs: <QuerySet [<CalibrationEvent: 08/29/2020>]>
Inv confconst {}
Part: CTDMO-Q (id=814)
Part calib fields     (REQUIRED): {'cal2'}
Part confconst fields (REQUIRED): {'const1'}
MODE: "part has confconsts"
  FAIL: inventory confconst is empty

Inv: CGINS-DOSTAD-00126 (id=1437)
Inv calibs: <QuerySet [<CalibrationEvent: 08/28/2020>]>
Inv confconst {}
Part: DOSTA-D (id=816)
Part calib fields     (REQUIRED): {'cal2', 'cal1'}
Part confconst fields (REQUIRED): {}
MODE: "part has calibs ONLY"
  <CalibrationEvent: 08/28/2020>
    calib approved:   FAIL
    calib validation: PASS

Inv: 1336-00001-00003-bugtest (id=3803)
Inv calibs: <QuerySet [<CalibrationEvent: 08/31/2020>, <CalibrationEvent: 08/28/2020>, <CalibrationEvent: 08/11/2020>]>
Inv confconst {34: <QuerySet [<ConfigEvent: 08/28/2020>, <ConfigEvent: 08/28/2020>]>}
Part: CTDBP-C (id=757)
Part calib fields     (REQUIRED): {'cal2', 'New2', 'New1', 'cal1'}
Part confconst fields (REQUIRED): {'new1'}
MODE: "part has confconsts"
  Deployment: "fixed - Testing At Sea" (id=34)
    conf/consts approved:  FAIL. Approvals={<ConfigEvent: 08/28/2020>: False, <ConfigEvent: 08/28/2020>: False}
    conf/const validation: PASS
    calib selected: <CalibrationEvent: 08/28/2020>
    calib approved:   PASS
    calib validation: PASS

Inv: PHSEN-E-Whaat? (id=3806)
Inv calibs: <QuerySet [<CalibrationEvent: 08/31/2020>, <CalibrationEvent: 08/12/2020>, <CalibrationEvent: 08/02/2020>, <CalibrationEvent: 08/01/2020>]>
Inv confconst {}
Part: PHSEN-E (id=823)
Part calib fields     (REQUIRED): {'again1'}
Part confconst fields (REQUIRED): {}
MODE: "part has calibs ONLY"
  <CalibrationEvent: 08/31/2020>
    calib approved:   FAIL
    calib validation: PASS
  <CalibrationEvent: 08/12/2020>
    calib approved:   FAIL
    calib validation: FAIL. Present=set() Missing={'again1'}
  <CalibrationEvent: 08/02/2020>
    calib approved:   FAIL
    calib validation: FAIL. Present=set() Missing={'again1'}
  <CalibrationEvent: 08/01/2020>
    calib approved:   FAIL
    calib validation: FAIL. Present=set() Missing={'again1'}

Inv: 3305-00012-00001-01329 (id=3808)
Inv calibs: <QuerySet [<CalibrationEvent: 09/01/2020>]>
Inv confconst {}
Part: PRESF-A (id=421)
Part calib fields     (REQUIRED): {'cal1'}
Part confconst fields (REQUIRED): {}
MODE: "part has calibs ONLY"
  <CalibrationEvent: 09/01/2020>
    calib approved:   FAIL
    calib validation: PASS

Inv: 3305-00012-00001-01330 (id=3809)
Inv calibs: <QuerySet [<CalibrationEvent: 09/01/2020>]>
Inv confconst {}
Part: PRESF-A (id=421)
Part calib fields     (REQUIRED): {'cal1'}
Part confconst fields (REQUIRED): {}
MODE: "part has calibs ONLY"
  <CalibrationEvent: 09/01/2020>
    calib approved:   FAIL
    calib validation: PASS

Inv: 666-01 (id=3810)
Inv calibs: <QuerySet [<CalibrationEvent: 09/14/2020>, <CalibrationEvent: 09/09/2020>]>
Inv confconst {41: <QuerySet [<ConfigEvent: 09/21/2020>, <ConfigEvent: 09/21/2020>]>}
Part: wozzle (id=1655)
Part calib fields     (REQUIRED): {'calA', 'calB'}
Part confconst fields (REQUIRED): {'constA', 'conf2'}
MODE: "part has confconsts"
  Deployment: "88.8 - Parking Lot" (id=41)
    conf/consts approved:  PASS
    conf/const validation: PASS
    calib selected: <CalibrationEvent: 09/14/2020>
    calib approved:   PASS
    calib validation: PASS
SUCCESS: 666-01__20200921.csv zipped!

Inv: CGINS-PHSENE-P0171 (id=2083)
Inv calibs: <QuerySet [<CalibrationEvent: 06/02/2020>]>
Inv confconst {}
FAIL: CGINS-PHSENE-P0171 does not have an associated assembly_part!

Inv: CGINS-CTDBPC-06841 (id=1129)
Inv calibs: <QuerySet [<CalibrationEvent: 07/15/2020>, <CalibrationEvent: 07/13/2020>]>
Inv confconst {}
FAIL: CGINS-CTDBPC-06841 does not have an associated assembly_part!

Inv: CGINS-NUTNRB-01094 (id=1966)
Inv calibs: <QuerySet [<CalibrationEvent: 01/28/2020>]>
Inv confconst {}
FAIL: CGINS-NUTNRB-01094 does not have an associated assembly_part!

Inv: CGINS-OPTAAD-00123 (id=1971)
Inv calibs: <QuerySet [<CalibrationEvent: 07/15/2020>]>
Inv confconst {}
FAIL: CGINS-OPTAAD-00123 does not have an associated assembly_part!

Inv: CGINS-CTDMOG-13419 (id=1333)
Inv calibs: <QuerySet [<CalibrationEvent: 02/22/2020>]>
Inv confconst {}
FAIL: CGINS-CTDMOG-13419 does not have an associated assembly_part!

Inv: CGINS-CTDMOG-13420 (id=1334)
Inv calibs: <QuerySet [<CalibrationEvent: 02/02/2020>]>
Inv confconst {}
FAIL: CGINS-CTDMOG-13420 does not have an associated assembly_part!

Inv: CGINS-CTDMOG-13431 (id=1343)
Inv calibs: <QuerySet [<CalibrationEvent: 08/29/2020>, <CalibrationEvent: 02/26/2020>]>
Inv confconst {}
Part: CTDMO-G (id=772)
Part calib fields     (REQUIRED): {'cal1'}
Part confconst fields (REQUIRED): {}
MODE: "part has calibs ONLY"
  <CalibrationEvent: 08/29/2020>
    calib approved:   FAIL
    calib validation: PASS
  <CalibrationEvent: 02/26/2020>
    calib approved:   FAIL
    calib validation: FAIL. Present={'CC_a1', 'CC_a0', 'CC_a3', 'CC_a2'} Missing={'cal1'}

```

**Comment by spetillo** [comment]
Looking good from what I can test right now!
